### PR TITLE
Add GSEA analysis

### PR DESCRIPTION
**Pull request by ajlee21**
This PR adds GSEA analysis.

The main changes can be found in `human_analysis/2_identify_generic_genes_pathways.ipynb`.

The goal of this notebook is to identify commonly DE genes and subsequent pathways. The following steps are performed in this notebook:
1. Simulates N gene expression experiments using ponyo
2. Perform DE analysis to get association statistics for each gene
3. For each gene, aggregate statsitics across all simulated experiments
4. Rank genes based on this aggregated statistic (i.e. log fold change, or p-value) where high ranked genes are those that are commonly DE.

Similar steps are performed to identify generic gene sets (pathways):

1. Using the same simulated experiments from above, perform GSEA analysis. This analysis will determine whether the genes contained in a gene set are clustered towards the beginning or the end of the ranked list of genes, where genes are ranked by say log fold change, indicating a correlation with change in expression.
2. For each gene set (pathway), aggregate statistics across all simulated experiments
3. Rank gene sets based on this aggregated statistic. Again, high rank refers to pathways that are significantly enriched frequently

We can then compare our ranking of genes and pathways vs those published. We find that our gene ranking is well correlated with published gene ranks. However, our ranking of pathways is not consistent and we're not sure why.

Additional minor changes were made to update supporting modules to perform gsea, update configs to use more user-friendly language for params, and the github actions test scripts but these don't need to be closely reviewed.

Note: Created an issue for the errors with github actions: https://github.com/greenelab/generic-expression-patterns/issues/16

### CODE REVIEW

**Review by danich1 — APPROVED** [review]
LGTM! Have a few comments/questions. Once done replying feel free to merge.

>Our results found ~3K DEGs which is close enough in range given that the data was processed using different methods. recount2 resource were aligned with the splice-aware Rail-RNA aligner

Did you found a large overlap between your 3k DEG and kim et al.'s 1459 DEG? Close enough on a couple thousand difference raises my statistical eyebrow; therefore, I wanted to confirm that I'm just a skeptical person.

>We suspect the reason for the p53 pathway not being significant in this case might be due to the small sample size here (6 paired samples). So for now we will move forward with our GSEA analysis.

Feel free to ignore suggestion: but is it possible to retry simulating with an experiment that's large enough to detect p53? Your results are great and glad to see it working, but I'm wondering how much push back you will get when it gets time for publication review. (Thinking down the road here 😄)

>Our top 20 pathways (out of a total of 50 pathways) are consistent with those found in Powers et. al. to be generic

When you said ^ did you mean that 20 pathways are consistent in terms of being enriched not consistent as in achieving the same rank right? Just confirming I'm reading your conclusion correctly.

File: human_analysis/nbconverted/2_identify_generic_genes_pathways.py [hunk 1]
```diff
@@ -113,12 +117,27 @@
...
+
+
 # ### Simulate experiments using selected template experiment
 
-# In[4]:
+# In[5]:
 
 
-# Simulate multiple experiments
+"""# Simulate multiple experiments
```
[the remark this answers is not in the document]
Comment (ajlee21) [reply human_analysis/nbconverted/2_identify_generic_genes_pathways.py:140]: These were cells that were commented out so I could test/debug without having to re-run my simulating experiments step which would take a little while. Need to uncomment these cells

File: human_analysis/nbconverted/2_identify_generic_genes_pathways.py [hunk 2]
```diff
@@ -171,32 +190,32 @@
...
 
 # Check ordering of sample ids is consistent between gene expression data and metadata
 process.compare_and_reorder_samples(template_data_file, metadata_file)
 
 
-# In[11]:
+# In[12]:
 
 
 get_ipython().run_cell_magic('R', '-i metadata_file -i project_id -i template_data_file -i local_dir', '\nsource(\'../generic_expression_patterns_modules/DE_analysis.R\')\n\nget_DE_stats_DESeq(metadata_file,\n             project_id, \n             template_data_file,\n             "template",\n             local_dir,\n             "real")')
```
[the remark this answers is not in the document]
Comment (ajlee21) [reply human_analysis/nbconverted/2_identify_generic_genes_pathways.py:215]: Yup..same..found this when there was one more R step i needed to add to my pipeline that was entirely in python!

**Comment by ajlee21** [comment]
> Did you found a large overlap between your 3k DEG and kim et al.'s 1459 DEG? Close enough on a couple thousand difference raises my statistical eyebrow; therefore, I wanted to confirm that I'm just a skeptical person.

This was initially my thought as well - that 1.5K to 3K was still on the large side. I was swayed by the quick manual spot check of DEGs and the fact that our processing of the data was different so I can't expect a very close overlap.

> Feel free to ignore suggestion: but is it possible to retry simulating with an experiment that's large enough to detect p53? Your results are great and glad to see it working, but I'm wondering how much push back you will get when it gets time for publication review. (Thinking down the road here 😄)

This is another good point. I did consider trying to simulate experiments with more than 6 samples, but this would require a bit of a refactor. Currently the way my method works is that it takes your chosen `template experiment` (real experiment), and saves the sample ids (which in this case there are 6). Then when it goes to simulate new simulated experiments, it selects the gene expression data using the sample ids (i.e. it queries my df of all gene expression data) then it encodes these samples into the VAE latent space and shifts it to produce a new simulated experiment. This process is repeated to get multiple simulated experiments. Since I have used the sample ids from the real experiment to simulate, I would need to adjust this process. This felt to me like a large enough undertaking that would cost more than it was worth to me. This wouldn't be a result figure of the eventual paper but was more a validation for me to make sure that my implementation was working.


> > Our top 20 pathways (out of a total of 50 pathways) are consistent with those found in Powers et. al. to be generic
> 
> When you said ^ did you mean that 20 pathways are consistent in terms of being enriched not consistent as in achieving the same rank right? Just confirming I'm reading your conclusion correctly.
 
Exactly! So the top 20 pathways found by our method were also the same ones that were found to be generic in Figure 4A of the paper. So there is consistency if we ignore the ordering of the pathways in terms of rank. But the issue is when we use the rank.

**Pull request merged**